nixos/modules/services/monitoring/dcgm.nix

```diff
@@ -0,0 +1,208 @@
+      defaultText = literalExpression ''
+        ''${config.services.dcgm.homeDir}/nv-hostengine.log"
+      '';
+      description = lib.mdDoc ''
+        Specify the filename nv-hostengine should use to dump logging information.
+      '';
+    };
+
+    logRotate = mkOption {
+      type = types.bool;
+      default = false;
+      description = lib.mdDoc ''
+        Rotate the log file if the log file with the same name already exists.
+      '';
+    };
+
+    denylistModules = mkOption {
+      type = types.nullOr types.commas;
+      default = null;
+      description = lib.mdDoc ''
+        DCGM modules that should be added to the Denylist so they aren't run by the hostengine.
+        Pass a comma-separated list of module IDs like 1,2,3.
+        Module IDs are available in dcgm_structs.h as DcgmModuleId constants.
+      '';
+    };
+
+    serviceAccount = mkOption {
+      type = types.str;
+      default = "dcgm";
+      description = lib.mdDoc ''
+        Service account that will be used for unprivileged processes.
+        Usually this should be set to the default nvidia-dcgm account.
+      '';
+    };
+
+    homeDir = mkOption {
+      type = types.path;
+      default = "/var/lib/dcgm";
+      description = lib.mdDoc ''
+        The full path to the directory that the DCGM diagnostic should be launched from.
+        This becomes the default output directory for DCGM diagnostic log and stats files.
+      '';
+    };
+
+    openFirewall = mkOption {
+      type = types.bool;
+      default = false;
+      description = lib.mdDoc ''
+        Open the firewall for the port.
+      '';
+    };
+  };
+
+  config = mkMerge [
+    (mkIf cfg.enable {
+      # TODO(@connorbaker): Look at
+      # https://github.com/nix-community/queued-build-hook/blob/6a69b97d2951e3440f209930cde0464d7754274c/module.nix
+      # for an example service which uses a socket.
+
+      # TODO(@connorbaker): Currently complaining about being run as non-root. We can't run it as the service account?
+
+      # TODO(@connorbaker): Current errors: (get new copy)
+
+      systemd.services.dcgm = {
+        description = "NVIDIA Data Center GPU Manager (DCGM)";
+        wantedBy = ["multi-user.target"];
+        after = ["network.target"];
+        environment = {
+          # NOTE: Add the libnvidia-ml.so.1 and libnvidia_nscq.so libraries to the LD_LIBRARY_PATH
+          # using the current NVIDIA driver.
+          # If we don't we get a failure here:
+          # https://github.com/NVIDIA/DCGM/blob/7e1012302679e4bb7496483b32dcffb56e528c92/sdk/nvidia/nvml/nvml_loader/nvml_loader.cpp#L199-L212.
+          # The underlying problem is that DCGM expects the library to be in certain paths:
+          # https://github.com/NVIDIA/DCGM/blob/7e1012302679e4bb7496483b32dcffb56e528c92/sdk/nvidia/nvml/nvml_loader/nvml_loader.cpp#L126-L131.
+          LD_LIBRARY_PATH = "${config.hardware.nvidia.package}/lib";
```

Comment:
`libnvidia-ml` has a hard dependency on the active kernel driver, so I'm pretty sure this will result in runtime errors if the driver gets updated.

Should probably use `addOpenGLRunpath.driverLink` instead.